library/src/commonMain/kotlin/com/badoo/reaktive/observable/Sample.kt

```diff
@@ -0,0 +1,54 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.disposable.CompositeDisposable
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.scheduler.Scheduler
+import com.badoo.reaktive.utils.lock.newLock
+import com.badoo.reaktive.utils.lock.synchronized
+
+fun <T> Observable<T>.sample(window: Long, scheduler: Scheduler): Observable<T> =
+    observableByEmitter { emitter ->
+        val disposableWrapper = CompositeDisposable()
+        emitter.setDisposable(disposableWrapper)
+        val executor = scheduler.newExecutor()
+        disposableWrapper += executor
+
+        subscribeSafe(
+            object : ObservableObserver<T> {
+                private val lock = newLock()
+                private lateinit var lastValue: MutableList<T>
```

Comment:
Why is this a list?

Reply:
We need some kind of mutable holder of nullable value so we can distinguish between the following cases:
1. There is no value at all (not initialised)
2. There is a value, null or not null.